{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1 read correct\n",
    "##### Quake下载和安装\n",
    "```bash\n",
    "wget http://www.cbcb.umd.edu/software/quake/downloads/quake-0.3.5.tar.gz\n",
    "tar -vxf quake-0.3.5.tar.gz \n",
    "cd Quake/src/\n",
    "make\n",
    "cd ../bin/\n",
    "./correct -h\n",
    "```\n",
    "##### Quake运行           Fail\n",
    "```bash\n",
    "python2 ~/software/Quake/bin/quake.py  -f filenames.tab -k 19 # filenames.tab保存fastq格式的原始文件路径，配对文件用空格格开\n",
    "./correct\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1 原始数据预处理"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1 原始数据质量控制\n",
    "```bash\n",
    "java -jar /home/wxu/software/Trimmomatic-0.36/trimmomatic-0.36.jar PE -phred33 -threads 24 \"../data/PMSuperPoolB_Y_R1.fastq.gz\" \"../data/PMSuperPoolB_Y_R2.fastq.gz\" \"paired_PMSuperPoolB_Y_R1.fastq.gz\" \"unpaired_PMSuperPoolB_Y_R1.fastq.gz\" \"paired_PMSuperPoolB_Y_R2.fastq.gz\" \"unpaired_PMSuperPoolB_Y_R2.fastq.gz\" HEADCROP:4 LEADING:20 TRAILING:20 ILLUMINACLIP:./adaptor.fasta:2:18:7:2:true SLIDINGWINDOW:5:20 AVGQUAL:23 MINLEN:70 \n",
    "\n",
    "\n",
    "fastqc -t 24 paired_PMSuperPoolB_Y_R1.fastq.gz\n",
    "\n",
    "fastqc -t 24 paired_PMSuperPoolB_Y_R2.fastq.gz\n",
    "\n",
    "\n",
    "java -jar /home/wxu/software/Trimmomatic-0.36/trimmomatic-0.36.jar PE -phred33 -threads 24 \"../data/PMSuperPoolB_X_R1.fastq.gz\" \"../data/PMSuperPoolB_X_R2.fastq.gz\" \"paired_PMSuperPoolB_X_R1.fastq.gz\" \"unpaired_PMSuperPoolB_X_R1.fastq.gz\" \"paired_PMSuperPoolB_X_R2.fastq.gz\" \"unpaired_PMSuperPoolB_X_R2.fastq.gz\" HEADCROP:4 LEADING:20 TRAILING:20 ILLUMINACLIP:./adaptor.fasta:2:18:7:2:true SLIDINGWINDOW:5:20 AVGQUAL:23 MINLEN:70\n",
    "\n",
    "\n",
    "fastqc -t 24 paired_PMSuperPoolB_X_R1.fastq.gz\n",
    "\n",
    "fastqc -t 24 paired_PMSuperPoolB_X_R2.fastq.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 污染序列去除（大肠杆菌和Phi174）\n",
    "```bash\n",
    "bowtie2 --threads 18 --un-conc ./CleanReads_PMSuperPoolB_X -x /home/wxu/data/genome/E_coli/ecoli_DH10B.fasta -1 ../trim/paired_PMSuperPoolB_X_R1.fastq.gz -2 ../trim/paired_PMSuperPoolB_X_R2.fastq.gz -S ./PM-B-pool-X-align.sam\n",
    "\n",
    "bowtie2 --threads 18 --un-conc ./CleanReads_PMSuperPoolB_Y -x /home/wxu/data/genome/E_coli/ecoli_DH10B.fasta -1 ../trim/paired_PMSuperPoolB_Y_R1.fastq.gz -2 ../trim/paired_PMSuperPoolB_Y_R2.fastq.gz -S ./PM-B-pool-Y-align.sam\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 数据拆分\n",
    "```bash\n",
    "# index.id文件的每行是每个混合池建库过程中使用的index序列\n",
    "~/software/bbmap/demuxbyname.sh in=/home/wxu/PMSuperPoolB/CleanReads/CleanReads_PMSuperPoolB_X.1 in2=/home/wxu/PMSuperPoolB/CleanReads/CleanReads_PMSuperPoolB_X.2 out=X_%_R1.fastq out2=X_%_R2.fastq names=index.id prefixmode=f\n",
    "\n",
    "~/software/bbmap/demuxbyname.sh in=/home/wxu/PMSuperPoolB/CleanReads/CleanReads_PMSuperPoolB_Y.1 in2=/home/wxu/PMSuperPoolB/CleanReads/CleanReads_PMSuperPoolB_Y.2 out=Y_%_R1.fastq out2=Y_%_R2.fastq names=index.id prefixmode=f\n",
    "\n",
    "# 输出的文件是以index序列命名的，需要用对应的混合池编号重命名\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2 混合池末端序列解析\n",
    "#### 2.1 短末端序列\n",
    "```bash\n",
    "# VES-BES reads提取\n",
    "python /home/wxu/workspace/BES-simulation/RICE/BES/BES_F_optimize/filtered_reads_from_NGS_20181107.py -v /home/wxu/data/vector/pIndigoBAC536S_F.fa -1 /home/wxu/workspace/PM-B-XY/data/${index}_R1.fastq -2 /home/wxu/workspace/PM-B-XY/data/${index}_R2.fastq -a ${index}_filtered_R1.fastq -b ${index}_filtered_R2.fastq\n",
    "# 文件格式转换和合并\n",
    "python /home/wxu/software/toolkits/fastq2fasta.py -i ../1_filtered_NGS/${index}_filtered_R1.fastq -o ${index}_R1.fasta\n",
    "python /home/wxu/software/toolkits/fastq2fasta.py -i ../1_filtered_NGS/${index}_filtered_R2.fastq -o ${index}_R2.fasta\n",
    "cat ${index}_R1.fasta ${index}_R2.fasta > ${index}.fasta\n",
    "\n",
    "rm ${index}_R1.fasta\n",
    "rm ${index}_R2.fasta\n",
    "\n",
    "# Assembled to Contig\n",
    "cap3 ${index}.fasta > ${index}.log\n",
    "\n",
    "rm ${index}.log\n",
    "rm ${index}.fasta.cap.info\n",
    "rm ${index}.fasta.cap.singlets\n",
    "rm ${index}.fasta.cap.ace\n",
    "rm ${index}.fasta.cap.contigs.qual\n",
    "rm ${index}.fasta.cap.contigs.links\n",
    "\n",
    "\n",
    "\n",
    "#scafolding with SCARPA\n",
    "/home/wxu/software/velvet_1.2.10/contrib/shuffleSequences_fasta/shuffleSequences_fastq.pl ../1_filtered_NGS/${index}_filtered_R1.fastq ../1_filtered_NGS/${index}_filtered_R2.fastq ${index}_R12.fastq\n",
    "/home/wxu/software/scarpa-0.241/bin/scarpa_process -c ${index}.fasta.cap.contigs -f ${index}_R12.fastq -i 500\n",
    "bowtie2-build ${index}.fasta.cap.contigs.scarpa.fa ${index}_index\n",
    "bowtie2 -x ${index}_index -U ${index}_R12.fastq.scarpa.fq | /home/wxu/software/scarpa-0.241/bin/scarpa_parser > ${index}_sam.scarpa\n",
    "/home/wxu/software/scarpa-0.241/bin/scarpa -c  ${index}.fasta.cap.contigs.scarpa.fa -i ${index}_sam.scarpa -l  ${index}.fasta.cap.contigs.scarpa.info -o ${index}.scaf.fasta\n",
    "\n",
    "rm ${index}*.bt2\n",
    "\n",
    "\n",
    "\n",
    "# Trimming Vector End Sequence\n",
    "blat /home/wxu/data/vector/pIndigoBAC536S_F.fa ${index}.scaf.fasta -noHead ${index}.psl\n",
    "python /home/wxu/software/toolkits/trim_vector_from_BES_by_blat_20181108.py -f ${index}.scaf.fasta -p ${index}.psl -o ${index}_BES.fasta\n",
    "```\n",
    "\n",
    "#### 2.2 长末端序列\n",
    "\n",
    "组装每个混合池序列\n",
    "\n",
    "```bash\n",
    "# index为混合池编号\n",
    "python ~/software/SPAdes-3.13.0-Linux/bin/spades.py -t 7 -k 31,51,71,91 --pe1-1 ../data/${index}_R1.fastq --pe1-2 ../data/${index}_R2.fastq -o ${index}\n",
    "\n",
    "cp ./${index}/contigs.fa ${index}_contigs.fa\n",
    "\n",
    "~/software/mashmap-Linux64-v2.0/mashmap -r ~/data/vector/BAC-contamination.fa -q ${index}/contigs.fasta --pi 95 -o ${index}/mashmap.out\n",
    "\n",
    "python decontainment_from_contigs_by_mashmap.20190219.py -p ${index}/mashmap.out -f ${index}/contigs.fasta -o ${index}_long_contigs.fa\n",
    "```\n",
    "\n",
    "从混合池组装结果中提取长末端序列\n",
    "\n",
    "~/software/toolkits/extract_aligned_sequence_20190314.py\n",
    "\n",
    "#### 2.3 末端序列注释\n",
    "```bash\n",
    "/home/hwang/biosoftware/RepeatMasker/RepeatMasker -parallel 20 -species maize -html -gff -dir long_BES_mask long_BES.fasta\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3 BAC末端序列交叉定孔"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```Python3\n",
    "import numpy as np\n",
    "from gresource import plate\n",
    "from gresource.align import TwoScaffoldsAlignment\n",
    "from gresourceold.plate import  flatten\n",
    "import HTSeq\n",
    "\n",
    "#layout_id = plate.parse_pool_identity(np.array([['1','2'],['3','4'],['5','6']]))\n",
    "layout_id = plate.superpool_2 #糜子384板的布局（25-48）\n",
    "\n",
    "\n",
    "from HTSeq import FastaReader\n",
    "import os\n",
    "\n",
    "\n",
    "handle_file_valid = open('BES_valid.fasta','w') #位点仅单条序列\n",
    "handle_file_invalid = open('BES_invalid.fasta', 'w') #位点有多条序列的情况\n",
    "handle_file = open('BES_ALL.fasta','w+')\n",
    "\n",
    "BES_library_0 = {} # initial dict\n",
    "\n",
    "num_bes = np.zeros((96,96), dtype='int8')  #建立空的矩阵，用于存储每个位点解析出的末端序列数目\n",
    "\n",
    "\n",
    "# 行列交叉比对，寻找成对的末端序列\n",
    "for i in range(layout_id.shape[0]):\n",
    "    print('Reading the %s' %i)\n",
    "    for j in range(layout_id.shape[1]):\n",
    "        query = j\n",
    "        target = i\n",
    "        BAC_id = layout_id[i,j]\n",
    "        os.system('python /home/wxu/software/toolkits/find_overlap_sequence_by_blat_20181112.py -q ../3_trimmed_pool/Y%s_BES.fasta -t ../3_trimmed_pool/X%s_BES.fasta -o BES_%s.fasta' % (query +1, target +1, BAC_id.decode()))\n",
    "\n",
    "        #read  BES to dict\n",
    "        if os.path.exists(str('BES_' + BAC_id.decode() + '.fasta')) == False:\n",
    "            os.system('touch BES_%s.fasta' % (BAC_id.decode()))\n",
    "        BES_library_0[BAC_id] = [ i for i in FastaReader(str('BES_' + BAC_id.decode() + '.fasta'))]\n",
    "\n",
    "\n",
    "\n",
    "# except false passitive\n",
    "def difference_set():\n",
    "#for times in range(5):                       # run five rounds\n",
    "    for i in range(layout_id.shape[0]):\n",
    "        print('Reading the %s' %i)\n",
    "        for j in range(layout_id.shape[1]):\n",
    "            this = layout_id[i, j]\n",
    "            if len(BES_library_0[this]) > 3 :    # only discuss multi BESs, the only one BES need to skip\n",
    "\n",
    "                X_ids_except_this = [x for x in layout_id[i, :] if x !=  this]\n",
    "                Y_ids_except_this = [y for y in layout_id[:, j] if y != this]\n",
    "\n",
    "                X_contigs_except_this = [k.name for k in flatten([BES_library_0[i][1::2] for i in X_ids_except_this])]\n",
    "                Y_contigs_except_this = [k.name for k in flatten([BES_library_0[i][::2] for i in Y_ids_except_this])]\n",
    "\n",
    "                del_index = []\n",
    "                for m in  range(int(len(BES_library_0[this])/2)):\n",
    "                    if BES_library_0[this][m*2 +1].name in X_contigs_except_this and BES_library_0[this][m*2].name in Y_contigs_except_this :\n",
    "                        del_index.append(m*2+1)\n",
    "                        del_index.append(m*2)\n",
    "                BES_library_0[this] = [v for index , v in enumerate(BES_library_0[this]) if index not in del_index]\n",
    "\n",
    "\n",
    "            num_bes[i, j] = int(len(BES_library_0[this]) / 2)  # count the number of BESs\n",
    "            with open('BES_'+ this.decode() + '_1.fasta', 'w') as handle:\n",
    "                for k in BES_library_0[this]:\n",
    "                    k.write_to_fasta_file(handle)\n",
    "    np.save('the_number_of_BESs', num_bes)\n",
    "\n",
    "#位点末端序列唯一性\n",
    "def only():\n",
    "    for i in range(layout_id.shape[0]):\n",
    "        for j in range(layout_id.shape[1]):\n",
    "            this = layout_id[i, j]\n",
    "            if len(BES_library_0[this]) == 2 :   # a pair of BESs\n",
    "\n",
    "                for row_id in [id for id in layout_id[i,:] if id != this]:\n",
    "                    if len(BES_library_0[row_id]) > 2:\n",
    "                        for x_index in range(len(BES_library_0[row_id])):\n",
    "                            if x_index % 2 == 0:\n",
    "                                continue\n",
    "                            else:\n",
    "                                print(i,j,this,row_id,x_index)\n",
    "                                if BES_library_0[this][1].name == BES_library_0[row_id][x_index].name:\n",
    "                                    del BES_library_0[row_id][x_index-1:x_index+1]\n",
    "                                    break\n",
    "\n",
    "                for column_id in [id for id in layout_id[:,j] if id != this]:\n",
    "                    if len(BES_library_0[column_id]) > 2:\n",
    "                        for y_index in range(len(BES_library_0[column_id])):\n",
    "                            if y_index % 2 == 0:\n",
    "                                if BES_library_0[this][0].name == BES_library_0[column_id][y_index].name:\n",
    "                                    del BES_library_0[column_id][y_index:y_index+2]\n",
    "                                    break\n",
    "\n",
    "\n",
    "            num_bes[i, j] = int(len(BES_library_0[this]) / 2)  # count the number of BESs\n",
    "            with open('BES_'+ this.decode() + '_1.fasta', 'w') as handle:\n",
    "                for k in BES_library_0[this]:\n",
    "                    k.write_to_fasta_file(handle)\n",
    "    np.save('the_number_of_BESs', num_bes)\n",
    "\n",
    "##chose a method to adjust BAC end sequence\n",
    "only()\n",
    "only()\n",
    "difference_set()\n",
    "only()\n",
    "\n",
    "#筛选后成对合并\n",
    "for i in range(layout_id.shape[0]):\n",
    "    for j in range(layout_id.shape[1]):\n",
    "        query = j\n",
    "        target = i\n",
    "        BAC_id = layout_id[i,j].decode()\n",
    "        print(BAC_id)\n",
    "        if os.path.exists('BES_%s_1.fasta' %(BAC_id)):\n",
    "            clone_list = [k for k in FastaReader('BES_%s_1.fasta' %(BAC_id))]\n",
    "\n",
    "            if len(clone_list) == 2 :\n",
    "                seq = TwoScaffoldsAlignment(SeqA=clone_list[0].seq.decode(), SeqB=clone_list[1].seq.decode()).merge()\n",
    "                HTSeq.Sequence(name=BAC_id, seq=seq.encode()).write_to_fasta_file(handle_file)\n",
    "                HTSeq.Sequence(name=BAC_id, seq=seq.encode()).write_to_fasta_file(handle_file_valid)\n",
    "            elif len(clone_list) > 2:  #若位点包含多条末端，在每条序列后面添加“_”和数字编号\n",
    "                for l in range(int(len(clone_list)/2)):\n",
    "                    seq = TwoScaffoldsAlignment(SeqA=clone_list[l*2].seq.decode(), SeqB=clone_list[l*2 + 1].seq.decode()).merge()\n",
    "                    name = BAC_id + '_'+ str(l+1)\n",
    "                    HTSeq.Sequence(name=name, seq=seq.encode()).write_to_fasta_file(handle_file)\n",
    "                    HTSeq.Sequence(name=name, seq=seq.encode()).write_to_fasta_file(handle_file_invalid)\n",
    "\n",
    "handle_file.close()\n",
    "handle_file_valid.close()\n",
    "handle_file_invalid.close()\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4 BAC定位"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.1 末端序列比对\n",
    "```bash\n",
    "blastn -db Panicum_miliaceum_longmi4.fna -query PM-B-BES.fasta -outfmt '6 qaccver saccver pident length mismatch gapopen qlen qstart qend slen sstart send evalue bitscore sstrand qcovhsp' -qcov_hsp_perc 99 -perc_identity 99 -lcase_masking > PM-B-BES_longmi4.unmasked.blast\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.2 GFF3格式形成"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```Python3\n",
    "class BAC_Recorder:\n",
    "    def __init__(self, record):\n",
    "        self._record = record\n",
    "        self._name = self._record.iloc[0, 16]\n",
    "        self._Forward = pd.DataFrame(columns=[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17])\n",
    "        self._Reverse = pd.DataFrame(columns=[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17])\n",
    "        self._gff = pd.DataFrame(columns=['chr','source','type','start','end','score','strand','phase','attributes'])\n",
    "\n",
    "\n",
    "        self.split()\n",
    "        self._Forward = self.merge(BES=self._Forward)\n",
    "        self._Reverse = self.merge(BES=self._Reverse)\n",
    "        self.conbinates()\n",
    "\n",
    "    def split(self):\n",
    "        for strand,BES_group in self._record.groupby([17]):\n",
    "            BES_group = BES_group.sort_values(by=[13],ascending=False)\n",
    "            if BES_group.iloc[0,17] == 'F':\n",
    "                self._Forward = pd.concat([BES_group,self._Forward])\n",
    "            elif BES_group.iloc[0,17] == 'R':\n",
    "                self._Reverse = pd.concat([BES_group,self._Reverse])\n",
    "\n",
    "    def merge(self, BES, edit=6):\n",
    "        num_rows = BES.shape[0]\n",
    "        BES[18] = -1\n",
    "        if num_rows :\n",
    "            for first in range(num_rows-1):\n",
    "                for second in range(first+1,num_rows):\n",
    "                    if BES.iloc[first,18] != -1 :\n",
    "                        continue\n",
    "                    else:\n",
    "                        if BES.iloc[first,1] == BES.iloc[second,1] and \\\n",
    "                        (abs(BES.iloc[first,10] - BES.iloc[second,10]) < edit or \\\n",
    "                         abs(BES.iloc[first,11] - BES.iloc[second,11]) < edit ):\n",
    "                            BES.iloc[first,18] = first\n",
    "                            BES.iloc[second,18] = first\n",
    "        return BES.drop_duplicates(subset=[18])\n",
    "\n",
    "    def conbinates(self, max_length=200000):\n",
    "        if self._Forward.shape[0] and self._Reverse.shape[0]:\n",
    "            FLAG = False\n",
    "            for F in range(self._Forward.shape[0]):\n",
    "                for R in range(self._Reverse.shape[0]):\n",
    "                    start = min(self._Forward.iloc[F, 10], self._Forward.iloc[F, 11], \\\n",
    "                                self._Reverse.iloc[R, 10], self._Reverse.iloc[R, 11])\n",
    "                    end   = max(self._Forward.iloc[F, 10], self._Forward.iloc[F, 11], \\\n",
    "                                self._Reverse.iloc[R, 10], self._Reverse.iloc[R, 11])\n",
    "                    if self._Forward.iloc[F, 1] == self._Reverse.iloc[R, 1] and \\\n",
    "                    self._Forward.iloc[F, 14] != self._Reverse.iloc[R, 14]  and \\\n",
    "                    end - start < max_length:\n",
    "                        self._gff.loc[self._name] = [self._Forward.iat[F,1],'GResource','BES', str(start), str(end), \\\n",
    "                                             '.','.','.',str('ID=' + self._name + ';Name=' + self._name)]\n",
    "                        self._gff.loc[str(self._name + '-F')] = [self._Forward.iat[F, 1], 'GResource', 'BES_end', \\\n",
    "                                    min(self._Forward.iat[F, 10], self._Forward.iat[F, 11]), \\\n",
    "                                    max(self._Forward.iat[F, 10], self._Forward.iat[F, 11]), \\\n",
    "                                    '.', self._Forward.iat[F, 14], '.', str('Parent=' + self._name)]\n",
    "                        self._gff.loc[str(self._name + '-R')] = [self._Reverse.iat[R, 1], 'GResource', 'BES_end', \\\n",
    "                                    min(self._Reverse.iat[R, 10], self._Reverse.iat[R, 11]), \\\n",
    "                                    max(self._Reverse.iat[R, 10], self._Reverse.iat[R, 11]), \\\n",
    "                                    '.', self._Reverse.iat[R, 14], '.', str('Parent=' + self._name)]\n",
    "                    elif self._Forward.iloc[F, 18] != -1:\n",
    "                        self._gff.loc[self._name] = [self._Forward.iat[F, 1], 'GResource', 'BES_end', \\\n",
    "                                             min(self._Forward.iat[F, 10], self._Forward.iat[F, 11]), \\\n",
    "                                             max(self._Forward.iat[F, 10], self._Forward.iat[F, 11]), \\\n",
    "                                             '.', self._Forward.iat[F, 14], '.', str('ID=' + self._name + ';Name=' + self._name)]\n",
    "                        FLAG = True\n",
    "                        break\n",
    "                    elif self._Reverse.iloc[R, 18] != -1:\n",
    "                        self._gff.loc[self._name] = [self._Reverse.iat[R, 1], 'GResource', 'BES_end', \\\n",
    "                                             min(self._Reverse.iat[R, 10], self._Reverse.iat[R, 11]), \\\n",
    "                                             max(self._Reverse.iat[R, 10], self._Reverse.iat[R, 11]), \\\n",
    "                                             '.', self._Reverse.iat[R, 14], '.', str('ID=' + self._name + ';Name=' + self._name)]\n",
    "                        FLAG = True\n",
    "                        break\n",
    "                if FLAG:\n",
    "                    break\n",
    "\n",
    "\n",
    "        elif self._Forward.shape[0]: # Only forward is recorded\n",
    "            for F in range(self._Forward.shape[0]):\n",
    "                self._gff.loc[self._name] = [self._Forward.iat[F, 1], 'GResource', 'BES_end', \\\n",
    "                                             min(self._Forward.iat[F, 10], self._Forward.iat[F, 11]), \\\n",
    "                                             max(self._Forward.iat[F, 10], self._Forward.iat[F, 11]), \\\n",
    "                                             '.', self._Forward.iat[F, 14], '.', str('ID=' + self._name + ';Name=' + self._name)]\n",
    "\n",
    "        elif self._Reverse.shape[0]:   # Only reverse BES is recorded\n",
    "            for R in range(self._Reverse.shape[0]):\n",
    "                self._gff.loc[self._name] = [self._Reverse.iat[R, 1], 'GResource', 'BES_end', \\\n",
    "                                             min(self._Reverse.iat[R, 10], self._Reverse.iat[R, 11]), \\\n",
    "                                             max(self._Reverse.iat[R, 10], self._Reverse.iat[R, 11]), \\\n",
    "                                             '.', self._Reverse.iat[R, 14], '.', str('ID=' + self._name + ';Name=' + self._name)]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "import collections\n",
    "\n",
    "\n",
    "def main(blast):\n",
    "    #add two field \"name\" and \"end\" into 17th and 18th respectively\n",
    "    for i in range(blast.shape[0]):\n",
    "        blast.loc[i, 16] = blast.loc[i, 0].split('_')[0]\n",
    "        blast.loc[i, 17] = blast.loc[i, 0].split('_')[-1]\n",
    "    blast[:3]\n",
    "\n",
    "    # create a new DataFrame with 9 columns of gff3 format\n",
    "    gff = pd.DataFrame(columns=['chr','source','type','start','end','score','strand','phase','attributes'])\n",
    "\n",
    "    # group by BAC name\n",
    "    for BES_name, BES_groups in blast.groupby(16):\n",
    "        BAC = BAC_Recorder(BES_groups)\n",
    "        gff = gff.append(BAC._gff)\n",
    "\n",
    "    return gff\n",
    "if __name__ == '__main__':\n",
    "    blast = pd.read_table('PM-B-BES_short_and_long.blast',header=None)\n",
    "    GFF3 = main(blast=blast)\n",
    "    GFF3.to_csv('PM-B-BES_20190407.gff3',index=None,sep='\\t')\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
